software/contrib/egressus_melodiam.py

```diff
@@ -0,0 +1,1091 @@
+            oled.pixel(15, 28, 1)
+            oled.pixel(15, 27, 1)
+            oled.pixel(15, 26, 1)
+            oled.pixel(16, 25, 1)
+            oled.pixel(16, 24, 1)
+            oled.pixel(16, 23, 1)
+
+        elif self.outputSlewModes[self.selectedOutput] == 3:  # expUpexpDown
+            oled.pixel(3, 31, 1)
+            oled.pixel(4, 30, 1)
+            oled.pixel(10, 30, 1)
+            oled.pixel(12, 30, 1)
+            oled.vline(5, 28, 2, 1)
+            oled.vline(9, 28, 2, 1)
+            oled.vline(13, 28, 2, 1)
+            oled.vline(6, 26, 2, 1)
+            oled.vline(8, 26, 2, 1)
+            oled.vline(14, 26, 2, 1)
+            oled.vline(7, 23, 3, 1)
+            oled.vline(15, 23, 3, 1)
+            oled.pixel(11, 31, 1)
+
+        elif self.outputSlewModes[self.selectedOutput] == 4:  # sharkTooth
+            oled.pixel(3, 30, 1)
+            oled.pixel(3, 29, 1)
+            oled.pixel(3, 28, 1)
+            oled.pixel(4, 27, 1)
+            oled.pixel(4, 26, 1)
+            oled.pixel(5, 25, 1)
+            oled.pixel(6, 25, 1)
+            oled.pixel(7, 24, 1)
+            oled.pixel(8, 24, 1)
+            oled.pixel(9, 23, 1)
+            oled.pixel(10, 24, 1)
+            oled.pixel(10, 25, 1)
+            oled.pixel(10, 26, 1)
+            oled.pixel(11, 27, 1)
+            oled.pixel(11, 28, 1)
+            oled.pixel(12, 28, 1)
+            oled.pixel(13, 29, 1)
+            oled.pixel(14, 29, 1)
+            oled.pixel(15, 30, 1)
+        elif self.outputSlewModes[self.selectedOutput] == 5:  # sharkToothReverse
+            oled.pixel(3, 30, 1)
+            oled.pixel(4, 29, 1)
+            oled.pixel(5, 29, 1)
+            oled.pixel(6, 28, 1)
+            oled.pixel(7, 28, 1)
+            oled.pixel(7, 27, 1)
+            oled.pixel(8, 26, 1)
+            oled.pixel(8, 25, 1)
+            oled.pixel(8, 24, 1)
+            oled.pixel(9, 23, 1)
+            oled.pixel(10, 24, 1)
+            oled.pixel(11, 24, 1)
+            oled.pixel(12, 25, 1)
+            oled.pixel(13, 25, 1)
+            oled.pixel(14, 26, 1)
+            oled.pixel(14, 27, 1)
+            oled.pixel(15, 28, 1)
+            oled.pixel(15, 29, 1)
+            oled.pixel(15, 30, 1)
+        elif self.outputSlewModes[self.selectedOutput] == 6:  # logUpStepDown
+            oled.vline(3, 28, 4, 1)
+            oled.hline(6, 24, 9, 1)
+            oled.pixel(4, 26, 1)
+            oled.pixel(4, 27, 1)
+            oled.pixel(5, 25, 1)
+            oled.vline(15, 24, 8, 1)
+        elif self.outputSlewModes[self.selectedOutput] == 7:  # stepUpExpDown
+            oled.hline(4, 24, 7, 1)
+            oled.vline(4, 24, 8, 1)
+            oled.pixel(10, 24, 1)
+            oled.pixel(10, 25, 1)
+            oled.pixel(10, 26, 1)
+            oled.pixel(11, 27, 1)
+            oled.pixel(11, 28, 1)
+            oled.pixel(12, 29, 1)
+            oled.pixel(12, 30, 1)
+            oled.pixel(13, 31, 1)
+            oled.pixel(14, 31, 1)

```

Comment:
[discussion]
Instead of having fixed `.pixel(...)` calls, which tend to be slow, it would probably be more efficient to define a byte array that can be used as a bitmap and drawn directly to the screen using `.blit(...)`. That would probably speed things up dramatically, as well as make the code a lot more readable.

Reply:
Great suggestion - thanks for the pointer. I have replaced all the pixel and line calls with a single call to a byte array for all wave shapes and the shred (new pattern generation) indicator. This has nicely improved performance.